Public/Src/Cache/ContentStore/Vfs/Provider/VfsProvider.cs

```diff
@@ -0,0 +1,507 @@
+            {
+                // The current process cannot trigger placeholder creation to prevent deadlock do to re-entrancy
+                // Just pretend the file doesn't exist.
+                return HResult.FileNotFound;
+            }
+
+            // FileRealizationMode.Copy = just create a normal placeholder
+
+            // FileRealizationMode.Hardlink:
+            // 1. Try to create a placeholder by hardlinking from local CAS
+            // 2. Create hardlink in VFS unified CAS
+
+            if (!Tree.TryGetNode(relativePath, out var node))
+            {
+                return HResult.FileNotFound;
+            }
+            else
+            {
+                relativePath = relativePath.EndsWith(node.Name) ? relativePath : Path.Combine(Path.GetDirectoryName(relativePath), node.Name);
+                if (node.IsDirectory)
+                {
+                    return virtualizationInstance.WritePlaceholderInfo(
+                        relativePath: relativePath,
+                        creationTime: node.Timestamp,
+                        lastAccessTime: node.Timestamp,
+                        lastWriteTime: node.Timestamp,
+                        changeTime: node.Timestamp,
+                        fileAttributes: node.IsDirectory ? FileAttributes.Directory : FileAttributes.Normal,
+                        endOfFile: node.Size,
+                        isDirectory: node.IsDirectory,
+                        contentId: s_contentId,
+                        providerId: s_providerId);
+                }
+                else
+                {
+                    return HandleCommandAsynchronously(commandId, async token =>
+                    {
+                        var fileNode = (VfsFileNode)node;
+                        await ContentManager.PlaceHydratedFileAsync(relativePath, new VfsFilePlacementData(fileNode.Hash, fileNode.RealizationMode, fileNode.AccessMode), token);
+
+                        // TODO: Create hardlink / move to original location to replace symlink?
+                        return HResult.Ok;
+                    });
+                }
+            }
+        }
+
+        internal HResult GetFileDataCallback(
+            int commandId,
+            string relativePath,
+            ulong byteOffset,
+            uint length,
+            Guid dataStreamId,
+            byte[] contentId,
+            byte[] providerId,
+            uint triggeringProcessId,
+            string triggeringProcessImageFileName)
+        {
+            // We should never create file placeholders so this should not be necessary
+            return HResult.FileNotFound;
+        }
+
+        private HResult QueryFileNameCallback(string relativePath)
+        {
+            // First try normal lookup
+            if (Tree.TryGetNode(relativePath, out var node))
+            {
+                return HResult.Ok;
+            }
+
+            if (!Utils.DoesNameContainWildCards(relativePath))

```

Comment:
>Utils.DoesNameContainWildCards [](start = 17, length = 30)

Do these play nicely with long paths?

Reply:
Don't know. This is implemented in the native layer. 

---
In reply to: [298428409](https://github.com/microsoft/BuildXL/pull/519#discussion_r298428409) [](ancestors = 298428409)